### ISSUE TITLE
How to use HELTEC WiFi Kit 32

### ISSUE THREAD

**Issue by mikekgr** [Edzelf/ESP32-Radio#112]
Dear Sirs,
I tried but unsuccessfully to work the HELTEC WiFi Kit 32 dev board with the ESP32-Radio. 
For your info, the attached pdf shows all the pins of this dev board, also the occupied pins regarding the I2C OLED ( that now supported in the ESP32-Radio thanks to Mr. @Edzelf  with his recent inclusion ).
If someone of you have this dev board working, or can suggest me what to do please send me a note.
All of my tries has always continious reset to ESP32... 
Thanks and Best Regards,
Mike Kranidis
[HELTEC WIFI_Kit_32-Diagram.pdf](https://github.com/Edzelf/ESP32-Radio/files/1959026/HELTEC.WIFI_Kit_32-Diagram.pdf)

**Comment by Edzelf** [comment]
First, try to run the radio software without anything attached (or configured) to the Heltec board.  If it does not crash, add the VS1063 and finally the OLED.  Remember to first run radio-init, so you can check the WIFi.

**Comment by mikekgr** [comment]
Dear @Edzelf ,
what I already tried is to put the correct I2C pins OLED_scl=15, OLED_sda=4, OLED_rst=16.
I successfully compiled then the "main" ino file ESP32_radio.ino and then I did power on/off to the HELTEC dev board. Looking at the serial port, it started and after a few seconds do reset again and again.
I am not sure regarding your suggestion: " Remember to first run radio-init, so you can check the WIFi. ".
What is that?
Again thanks and Best Regards,
Mike Kranidis

**Comment by Edzelf** [comment]
I also have a Heltec board.  I tried it now and it keeps rebooting.  I will see what's wronfg with this board.

**Comment by mikekgr** [comment]
Dear @Edzelf ,
yes I did, I fount Esp32_radio_init.ino inside the subdirectory tools and I loaded to Arduino IDE. Then after changed the lines 39,40: 
`preferences.putString ( "wifi_00", "myHomeWiFi/12345678" ) |
 preferences.putString ( "wifi_01", "SecondWiFi/12345678" ) |`
I compiled and sent to ESP32. At the serial I took "ESP32_radio_init completed...".
Then I loaded the Esp32_radio.ino to Arduino IDE and the result is above, keep resetting...
Thanks

**Comment by Edzelf** [comment]
The program crashed when no IR was configured.  I have made a correction in the new version.

**Comment by mikekgr** [comment]
Thanks a lot dear @Edzelf  for your fast responce and the bug fixing!
I will try it when at home.
Have a nice day.

**Comment by mikekgr** [comment]
Dear @Edzelf ,
I have the HELTEC board with me so I can tested in my workplace and I did.
I discovered that If I change the "default" display from BLUETFT to OLED then I still get the "always restarting and crashing. If I leave your default BLUETFT then it is starting without problems.
Please check if you can.
Thanks
```
//#define BLUETFT                      // Works also for RED TFT 128x160
#define OLED                         // 64x128 I2C OLED
```

**Comment by Edzelf** [comment]
I could not reproduce that.  I made a small correction.  The display will work now without VS1063.

**Comment by mikekgr** [comment]
dear @Edzelf  making the diff between your latest Esp32_radio.ino and the Esp32_radio.ino that was 5 days ago I found them 100% the same. Can you check and make sure that the today changes has been uploaded to github? Thanks

**Comment by Edzelf** [comment]
Yeah, wrong directory

**Comment by mikekgr** [comment]
Dear @Edzelf ,
yes it is working now. To help somebody looking the same, to have OLED activated in ESP32-Radio using HELTEC WiFi Kit 32 { ESP32 }, the following changes should be done to .ino file:
```
change / correct the next two lines 2380, 2381
    { "pin_tft_scl",  &ini_block.tft_scl_pin,     15          },   // Display I2C version /// by mikek ///
    { "pin_tft_sda",  &ini_block.tft_sda_pin,     4          },   // Display I2C version /// by mikek ///

/// put these lines just bellow Serial.begin ( 115200 ) ;                              // For debug   in void setup()
  /// by mikek ///
  pinMode(16,OUTPUT); /// THIS IS THE RESET PIN OF HELTEC OLED ... ///
  digitalWrite(16, LOW);    // set GPIO16 low to reset OLED
  delay(50); 
  digitalWrite(16, HIGH); // while OLED is running, must set GPIO16 in high   
```
The only small glitch remained and that is: the OLED does not show correctly the last (?) line, I see only some pixels, but this is trivial.

Thanks a lot Mr. @Edzelf  for your help and your fantastic work in ESP32-Radio !!!

**Comment by Edzelf** [comment]
Last line on OLED: must be a long text line in the lower section.  Try to change the volume, this will paint the last pixel line of the display.
And try to run th OLED without using pin 16.  I don't think the OLED needs a reset.

**Comment by mikekgr** [comment]
Dear @ Pin 16 must be High in order to get OLED up and running. if you don't initialize then OLED is off.

**Comment by Edzelf** [comment]
Yes, I understand.  But 16 will be pulled to high by the software.  I think it is not necessary to pulse it low for reset.

**Comment by mikekgr** [comment]
Dear @Edzelf  I am not sure how to connect VS1053 to HELTEC Dev board. Can you suggesting me the pin to pin connections?
Thanks.

**Comment by Edzelf** [comment]
As I said: Gpio 16 is pulled up by the software.  All unused pins are pulled HIGH.  You will see it in the debug logging: "D:GPIO16 is HIGH".

**Comment by mikekgr** [comment]
Yes you are right, I can confirm @Edzelf  that the lines I put at void setup to set GPIO 16 not needed. Thanks

**Comment by Edzelf** [comment]
To connect the VS1053 to the Heltec board, you may us someting like:
```
// Heltec   Signal  Wired to OLED       Wired to VS1053      SDCARD   Wired to the rest
// -------- ------  --------------      -------------------  ------   ---------------
// GPIO16           Reset               -                     -       -
// GPIO5            -                   pin 2 XCS             -       -
// GPIO4            SDA                 -                     -       -
// GPIO2            -                   pin 4 DREQ            -       -
// GPIO17           -                   -                     CS      -
// GPIO18   SCK     -                   pin 5 SCK             CLK     -
// GPIO19   MISO    -                   pin 7 MISO            MISO    -
// GPIO23   MOSI    -                   pin 6 MOSI            MOSI    -
// GPIO15           SCL                 -                     -       -
// GPI03    RXD0    -                   -                     -       Reserved serial input
// GPIO1    TXD0    -                   -                     -       Reserved serial output
// GPIO34   -       -                   pin 1 XDCS            -       Optional pull-up resistor
// GPIO35   -       -                   -                     -       Infrared receiver VS1838B
// GPIO25   -       -                   -                     -       Rotary encoder CLK
// GPIO26   -       -                   -                     -       Rotary encoder DT
// GPIO27   -       -                   -                     -       Rotary encoder SW
// -------  ------  ---------------     -------------------  ------   ----------------
// GND      -       pin 8 GND           pin 8 GND                     Power supply GND
// VCC 5 V  -       pin 7 BL            -                             Power supply
// VCC 5 V  -       pin 6 VCC           pin 9 5V                      Power supply
// EN       -       pin 1 RST           pin 3 XRST                    -

```

**Comment by mikekgr** [comment]
Dear @Edzelf  
My VS1053 has the following pinout:

[VS1053 Module](https://www.aliexpress.com/item/VS1053-VS1053B-MP3-Module-For-Arduino-UNO-Breakout-Board-With-SD-Card-Slot-VS1053B-Ogg-Real/32816427942.html?spm=a2g0s.9042311.0.0.aKUqEs
)
Pin.1  +5V
Pin.2  GND
Pin.3  CS
Pin.4  MISO
Pin.5  MOSI
Pin.6  SCK
Pin.7  XCS
Pin.8  XRESET
Pin.9  XDCS
Pin.10 DEREQ

following your proposed pinout I have two questions, 
1. What signal is your designated signal ( at HELTEC side ) EN that is connected with XRST of VS1053? My HELTEC board does not have it. I presume that you speak for XRESET as it been referenced to my VS1053 board
2. My VS1053 board has also CS. Where to connect ( if any ) ?

Thanks a lot

**Comment by Edzelf** [comment]
1 - RST on your Heltec board is the same as EN.
2 - CS is probably for the SD card, so you can use GPIO17 for this.

**Comment by mikekgr** [comment]
Dear @Edzelf  Many thanks. It is working fine now!

**Comment by EvgenPavlyuchek** [comment]
i have same HELTEC WiFi Kit 32
thx, it is ok with oled now
but nothing with sound (
i have such blue VS1053
https://ru.aliexpress.com/item/VS1053-MP3-w-On-Board-SPI/32839832382.html?spm=a2g0v.search0604.3.8.27af7892xcHZ9f&ws_ab_test=searchweb0_0%2Csearchweb201602_4_10065_10068_319_10059_10884_317_10887_10696_100031_321_322_10084_453_10083_454_10103_10618_10307_537_536%2Csearchweb201603_51%2CppcSwitch_0&algo_pvid=bb65d642-29d3-4045-a598-38d92f52237c&algo_expid=bb65d642-29d3-4045-a598-38d92f52237c-1
but after connection I hear nothing
how to find the reason ?

**Comment by FrankBoesing** [comment]
I had to insert a manual OLED reset to make the display work:

SSD1306.h, first lines of constructor:
`//Reset OLED
pinMode(16, OUTPUT);
digitalWrite(16, LOW);
delay(50);
digitalWrite(16, HIGH);
delay(50);
`

**Comment by timo619** [comment]
Hi, I have an Heltec ESP32 WiFi Lora Board [https://www.fambach.net/esp32-wifi-lora-433/](https://www.fambach.net/esp32-wifi-lora-433/). Therefore the pin layout is different. I have a working PCM5102A and a rotary encoder connected to it. 

The OLED display is not working yet.

Before configuring the GPIO I have the error:
`E (942333) i2c: i2c_set_pin(860): scl and sda gpio numbers are the same`

After adding the OLED GPIO (SCL: 15 and SDA: 4) I get the following error (audio and encoder still works), but not OLED display:
```
D: pin_enc_clk set to 17
D: pin_enc_dt set to 22
D: pin_enc_sw set to 2
D: pin_tft_scl set to 15
D: pin_tft_sda set to 4
D: pin_i2s_bck set to 25
D: pin_i2s_lck set to 14
D: pin_i2s_din set to 27
D: pin_spi_sck set to 18
D: pin_spi_miso set to 19
D: pin_spi_mosi set to 23
```
```
D: GPIO0 is HIGH
D: GPIO2 is HIGH
D: GPIO4 is HIGH
D: GPIO5 is HIGH
D: GPIO12 is HIGH
D: GPIO13 is HIGH
D: GPIO14 is HIGH
D: GPIO15 is HIGH
D: GPIO16 is HIGH
D: GPIO17 is LOW, probably no PULL-UP
D: GPIO18 is HIGH
D: GPIO19 is HIGH
D: GPIO21 is HIGH
D: GPIO22 is LOW, probably no PULL-UP
D: GPIO23 is HIGH
D: GPIO25 is HIGH
D: GPIO26 is LOW, probably no PULL-UP
D: GPIO27 is HIGH
D: GPIO32 is LOW, probably no PULL-UP
D: GPIO33 is LOW, probably no PULL-UP
D: GPIO34 is LOW, probably no PULL-UP
D: GPIO35 is LOW, probably no PULL-UP
D: GPIO39 is LOW, probabl���������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������
```
Any suggestion where the problem comes from? (tested the display and gpios with Adafruit GFX library)

**Comment by timo619** [comment]
The cause is a problem with the Wire.cpp and ESP32 and Platformio. My PlatformIO setup apparently cannot use the Wire.cpp with the ESP32. Simple GFXSketch shows the error: Bus is in Slave Mode. Same GFX script in Arduino IDE causes no problems and OLED works.

Any suggestions which setting/library mismatch causes the error and therefore non-working OLED on Platformio?

**Comment by Edzelf** [comment]
Maybe not the cause, but your 3 rotary encoder pins should be high at start-up.

**Comment by Edzelf** [comment]
Try to disconnect the display.  It may consume too much power for your power supply.
Try version 2 of the radio software.

**Comment by timo619** [comment]
I am testing with a simple code snippet with only i2c scanner + OLED, compiles on both (Arduino IDE and PlatformIO), but only works on the Heltec Dev Board, when compiled on Arduino IDE. Therefore I assume, that the cause has to come from my settings/libraries in PlatformIO.

main.cpp (OLED+i2c works, when compiled with Arduino IDE)
```
#include "Arduino.h"
#include "heltec.h"

void setup()
{
	Heltec.begin(true, false, true);
	Wire.begin(SDA_OLED, SCL_OLED); //Scan OLED's I2C address via I2C0
	//Wire1.begin(SDA, SCL);        //If there have other device on I2C1, scan the device address via I2C1
}

void loop()
{
	byte error, address;
	int nDevices;

	Serial.println("Scanning...");

	nDevices = 0;
	for(address = 1; address < 127; address++ )
	{
		Wire.beginTransmission(address);
		error = Wire.endTransmission();

//		Wire1.beginTransmission(address);
//		error = Wire1.endTransmission();

		if (error == 0)
		{
			Serial.print("I2C device found at address 0x");
			if (address<16)
			Serial.print("0");
			Serial.print(address,HEX);
			Serial.println("  !");

			nDevices++;
		}
		else if (error==4)
		{
			Serial.print("Unknown error at address 0x");
			if (address<16)
				Serial.print("0");
			Serial.println(address,HEX);
		}
	}
	if (nDevices == 0)
	Serial.println("No I2C devices found\n");
	else
	Serial.println("done\n");

	delay(5000);
}
```
platformio.ini
```
[env:heltec_wifi_lora_32_Test2]
platform = espressif32
board = heltec_wifi_lora_32
framework = arduino
monitor_speed = 115200
monitor_flags = 
	--filter=esp32_exception_decoder
	--echo
build_flags = 
	-DCORE_DEBUG_LEVEL=0
	-Os
lib_deps = 
	heltecautomation/Heltec ESP32 Dev-Boards @ ^1.1.0
```

**Comment by timo619** [comment]
I am using Version 2 of your script and I could narrow my problem down to the following code line, which is not executed correctly:

oled.cpp in the OLED constructor
`i2c_param_config ( I2C_NUM_0, &i2c_config ) ;`

Unfortunately my coding knowledge stops there. I also don't see any CLK or DATA on the i2c-lines with my oscilloscope. Which makes sense, since the i2c is not started correctly, in my opinion.

The display works with other libraries with PlatformIO. Therefore:
- display connected correct
- enough power supplied
- i2c is working with other libraries (Wire.h, Adafruit GFX)
- tested with and without encoder
- tested with second bus `i2c_param_config ( I2C_NUM_1, &i2c_config ) ;`
- tested slower i2c speed: 100000
- all other functionalities work: i2s (PCM5102A), WLAN, webserver, rotary encoder

Any further tips?

**Comment by Edzelf** [comment]
The I2C address of the OLED is 0x3C.  Is that equal to the adress of you OLED?
You have to define OLED1309, OLED1306 or OLED1106 in confog.h.Maybe you can add a simple I2Cscan function in setup().
The SDA and SCL pin numbers are printen in the serial monitor.  Are they correct for your set-up?

**Comment by timo619** [comment]
- My OLED address is also 0x3C
- I defined the OLED (#define OLED1306) in config.h (just the one)
- the pin numbers are printed correct (D: pin_tft_scl set to 15, D: pin_tft_sda set to 4)
- I cannot establish a i2c connection, since: (`i2c_param_config ( I2C_NUM_0, &i2c_config ) ;`) is not working
- therefore I think, that a scanner would not work and additionally I cannot program one
- the i2c scanner examples Iuesd all use the Wire.h library and with that the OLED is found correctly

**Comment by Edzelf** [comment]
There is a chance that there is output to the display before proper initialization.  I will check that.

**Comment by Edzelf** [comment]
I have changed the OLED library.  It will now use the Wire library.  You also have to replace the main program.
Please test it.

**Comment by timo619** [comment]
Thank you very much for your effort. 
I tested the up-to-date version (v2 / 04May2022) with the Wire library. No errors, but also no signals on the i2c sda and clk lines. Will test it further with i2c scanner the next days. Probably a weird behaviour of the specific heltec esp32 lora oled board.

**Comment by Edzelf** [comment]
There may be a reset pin for the display.  Try to pull this high.

**Comment by Edzelf** [comment]
Good!  But strange.....

**Comment by timo619** [comment]
I probably found the real cause of the problem:
[https://esp32.com/viewtopic.php?t=21030](https://esp32.com/viewtopic.php?t=21030 )

There was a new item added to the i2c config struct. I had to add the line: 

`i2c_config.clk_flags = 0 ; `

to the constructor. Then the i2c works with your library. Thanks again.

**Comment by frycol** [comment]
> I probably found the real cause of the problem: https://esp32.com/viewtopic.php?t=21030
> 
> There was a new item added to the i2c config struct. I had to add the line:
> 
> `i2c_config.clk_flags = 0 ; `
> 
> to the constructor. Then the i2c works with your library. Thanks again.

Can You tell me where exactly you add "i2c_config.clk_flags = 0 ;"

**Comment by Edzelf** [comment]
A crash like that is usually caused by a weqak power supply.

**Comment by Codemixer271** [comment]
Thanks for your answer.
I can exclude a week power supply. (tested the circuit with a laboratory power supply)
The hardware was working for about 2 years without any issue.
The ESP32 got some problems to connect to the network sometimes, so I decided to rebuild the software.
When I build the firmware without a display (activated DUMMYTFT), all is OK.
Building with each one of the OLED options results in this debug, display stays dark, but the radio is working.

Meanwhile I also tried the code on a WEMOS 32, without anything connected  and it runs into the same  problem.

**Comment by Edzelf** [comment]
Could you please try the version 2 of the radio?

**Comment by Codemixer271** [comment]
Great !
Was a good Christmas gift.
With V2 it´s running.

Thanks a lot for your support.
